cpp/ql/lib/semmle/code/cpp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -12,6 +12,14 @@ private import semmle.code.cpp.dataflow.ExternalFlow
 private import semmle.code.cpp.ir.IR
 
 module Input implements InputSig<DataFlowImplSpecific::CppDataFlow> {
+  /**
+   * Gets a string representing a level of indirection, for example for
+   * `indirection = 2`, the result is `**`.
+   */
+  private bindingset[indirection] string indirectionString(int indirection) {
+    result = concat(int i | i in [1 .. indirection] | "*")
+  }
```

Comment:
Yep, I was going to point this out later as well 😄 It's not clear to me whether the `concat` approach is faster or slower than the `prefix` approach I've added in `DataFlowUtil`. So some experimentation will be necessary there. I have a feeling that the `prefix` approach is faster, but I'm not sure

Reply:
Done - I've unified the four versions and made it clear than `n` can indeed be zero.

Note that the three older versions of this predicate used recursion and an upper bound, rather than `concat` and a `bindingset` annotation.  Performance characteristics will be different.  I'm expecting them to be better now (because `concat` appears to be much faster for all but huge strings), but the situation is complicated enough I will defer to the results of a DCA run (coming some time this week) for confirmation.

Reply:
> I'm expecting them to be better now (because concat appears to be much faster for all but huge strings), but the situation is complicated enough I will defer to the results of a DCA run (coming some time this week) for confirmation.

This surprises me. The `concat` solution will involve running a HOP, and because there's a bindingset annotation it will be inlined into every call site, meaning that the HOP will be invoked many many times. On the other hand the `prefix` solution will compute a very small table containing stars, and do a join with that.

With that said, I'm quite intrigued by this question. So I think I'll experiment with this a bit.

Reply:
I encountered a `prefix`-based solution as well - `"**********".prefix(indirection)`.  It's terrible because of that arbitrary length limit, but in practice it's probably really fast and plenty robust enough.  It is tempting.

Reply:
Indeed. That's what I used to do for local debugging. I think we can delay this choice until some later time, though. As long as we have a single place where all of this is computed (which you've ensured in https://github.com/github/codeql/pull/15371/commits/76780d74d96e754a18cdfd98b03962c622f95a99) I'm happy for now.